src/canard.c

```diff
@@ -1,6 +1,736 @@
+    uint32_t rhs_arb11 = rhs_ext ? (rhs_clean_id >> 18) : rhs_clean_id;
+    if (arb11 != rhs_arb11)
+    {
+        return arb11 < rhs_arb11;
+    }
+    else
+    {
+        return rhs_ext;
+    }
+  }
+
+  /*
+   * RTR vs Data frame - if frame identifiers and frame types are the same, RTR loses.
+   */
+  bool rtr     = id     & CANARD_CAN_FRAME_RTR;
+  bool rhs_rtr = rhs & CANARD_CAN_FRAME_RTR;
+  if (clean_id == rhs_clean_id && rtr != rhs_rtr)
+  {
+    return rhs_rtr;
+  }
+
+  /*
+   * Plain ID arbitration - greater value loses.
+   */
+  return clean_id < rhs_clean_id;
+}
+
+/**
+ * preps the rx state for the next transfer. does not delete the state
+ */
+CANARD_INTERNAL void canardPrepareForNextTransfer(CanardRxState* state)
+{
+  state->buffer_blocks = NULL;  //payload should be empty anyway
+  state->transfer_id += 1;
+  if (state->transfer_id >= 32)
+  {
+    state->transfer_id = 0;
+  }
+  state->payload_len = 0;
+  state->next_toggle = 0;
+  return;
+}
+
+/**
+ * returns data type from id
+ */
+CANARD_INTERNAL uint16_t canardDataType(uint32_t id)
+{
+  uint8_t transfer_type = canardTransferType(id);
+  if (transfer_type == CanardTransferTypeBroadcast)
+  {
+    return (uint16_t)CANARD_MSG_TYPE_FROM_ID(id);
+  } else
+  {
+    return (uint16_t)CANARD_SRV_TYPE_FROM_ID(id);
+  }
+}
+
+/**
+ * returns transfer type from id
+ */
+CANARD_INTERNAL uint8_t canardTransferType(uint32_t id)
+{
+  uint8_t is_service= (uint8_t)CANARD_SERVICE_NOT_MSG_FROM_ID(id);
+  if (is_service == 0)
+  {
+    return CanardTransferTypeBroadcast;
+  }
+  else if (CANARD_REQUEST_NOT_RESPONSE_FROM_ID(id) == 1)
+  {
+    return CanardTransferTypeRequest;
+  }
+  else
+  {
+    return CanardTransferTypeResponse;
+  }
+}
+
+/**
+ *  CanardRxState functions
+ */
+
+/**
+ * Traverses the list of CanardRxState's and returns a pointer to the CanardRxState 
+ * with either the Id or a new one at the end
+ */
+CANARD_INTERNAL CanardRxState *canardRxStateTraversal(CanardInstance* ins, uint32_t transfer_descriptor)
+{
+  CanardRxState* states = ins->rx_states;
+
+  if (states==NULL)
+  {           //initialize CanardRxStates
+    states = canardCreateRxState(&ins->allocator, transfer_descriptor);
+    ins->rx_states = states;
+    return states;
+  }
+
+  states = canardFindRxState(states, transfer_descriptor);
+  if (states != NULL)
+  {
+    return states;
+  }
+  else
+  {
+    return canardAppendRxState(ins, transfer_descriptor);
+  }
+}
+
+/**
+ * returns pointer to the rx state of transfer descriptor or null if not found
+ */
+CANARD_INTERNAL CanardRxState *canardFindRxState(CanardRxState* state, uint32_t transfer_descriptor)
+{
+  while(state != NULL)
+  {
+    if(state->dtid_tt_snid_dnid == transfer_descriptor)
+    {
+      return state;
+    }
+    state = state->next;
+  }
+  return NULL;
+}
+
+/**
+ * appends rx state to the canard instance rx_states
+ */
+CANARD_INTERNAL CanardRxState *canardAppendRxState(CanardInstance* ins, uint32_t transfer_descriptor)
```

Comment:
Appending to the end of the linked list is computationally expensive as it requires traversal. Prepending the list, however, is constant time.

```c
// This snippet has not been tested!
CanardRxState* new_state = canardCreateRxState(&ins->allocator, transfer_descriptor);
new_state->next = ins->rx_states;
ins->rx_states = new_state->next;
```